### DISCUSSION TITLE
Unusual behavior in the graphs resulting from model.train

### DISCUSSION THREAD

**Discussion by Josephts1** [ultralytics/ultralytics#18669] [Q&A] [answered]
Good morning, kind regards.
I am using YOLO for the classification of a class (fruits). I have made my own dataset with training (80 images), validation (15 images) and testing (10 images) data. When applying the attached code and reviewing the results returned by model.train (see attached image), I notice unusual behavior in these plots, such as sudden variations in the val/cls_loss, metrics/precision(B), metrics/recall(B), metrics/mAP50(B) or metrics/mAP50-95(B) plots. I have obtained similar results with YOLO versions 10 and 11 and tried to freeze the YOLO pre-trained weights with the COCO dataset.
I want to eliminate those large variations and have a properly exponential workout.
Thank you very much, I appreciate your knowledgeable input.
`from google.colab import drive
drive.mount('/content/drive')

import yaml

data={
    'path': '/content/drive/MyDrive/Proyecto_de_grado/data',
    'train': 'train',
    'val': 'val',
    'names': {
        0: 'fruta'
    }
}

with open('/content/drive/MyDrive/Proyecto_de_grado/data.yaml', 'w') as file:
    yaml.dump(data, file,default_flow_style=False,sort_keys=False)

!pip install ultralytics

from ultralytics import YOLO

model=YOLO('yolo11s.pt')
#CONGELAR CAPAS
Frez_layers=24 #Cantidad de capas a congelar máx 23. Capas backbone hasta la 9. Capas neck de la 10 a la 22.
freeze = [f"model.{x}." for x in range(0,Frez_layers)]  # capas "module" congeladas
print(freeze)
frozen_params={}
for k, v in model.named_parameters():
  #print(k)
  v.requires_grad = True  # train all layers
  frozen_params[k] = v.data.clone()
  #print(v.data.clone())
  #print(v.requires_grad)
  if any(x in k for x in freeze): #Si uno de los elementos en freeze es una subcadena del texto k, entra al bucle
   print(f"freezing {k}")
   v.requires_grad = False

result=model.train(data="/content/drive/MyDrive/Proyecto_de_grado/data.yaml",
                   epochs=100,patience=50,batch=16,plots=True,optimizer="auto",lr0=1e-4,seed=42,project="/content/drive/MyDrive/Proyecto_de_grado/runs/freeze_layers/todo_congelado_11s")

metrics = model.val(data='/content/drive/MyDrive/Proyecto_de_grado/data.yaml',
                    project='/content/drive/MyDrive/Proyecto_de_grado/runs/validation/todo_congelado_11s')
print(metrics)
print(metrics.box.map) #mAP50-95`
![Captura de pantalla 2025-01-12 082813](https://github.com/user-attachments/assets/26ad086a-a528-402a-be26-120f56558401)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Josephts1, thank you for sharing your question and for your interest in Ultralytics 🚀! We appreciate the detailed explanation and the inclusion of your training setup and plots—it’s super helpful!

If this is a 🐛 Bug Report, could you provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/) along with specific output logs or training artifacts? This will enable us to pinpoint the issue more effectively. If you're encountering unusual behavior in the plots, a closer look at your dataset, model configuration, or training logs would be helpful. 

For customized training setup ❓ Questions like this, we recommend verifying that your configuration aligns with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Diverse batch sizes, dataset checks, and possibly tweaking the `frozen layer` settings may play a factor here 🧠.

Meanwhile, let's also ensure you are running the latest version of the `ultralytics` package. Upgrade in your Python environment with the following command to rule out any resolved issues:

```bash
pip install -U ultralytics
```

Verify that your runtime environment satisfies the requirements, including [**Python>=3.8**](https://www.python.org/) and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). YOLO can run on multiple verified environments, such as:

- **Free GPUs in Notebooks:** <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud Platforms:** [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) and [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Containers:** See our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For further assistance:
- Explore our [Docs](https://docs.ultralytics.com) if you are new to YOLO, where you'll find [Python Usage Examples](https://docs.ultralytics.com/usage/python/) and [CLI examples](https://docs.ultralytics.com/usage/cli/).
- Connect with our vibrant user community on [Discord](https://discord.com/invite/ultralytics) 🎧, [Discourse](https://community.ultralytics.com), or [Reddit](https://reddit.com/r/Ultralytics) to exchange ideas and solutions.

Lastly, rest assured, this is an automated response 🤖, but an Ultralytics engineer will review your question shortly and provide insights to guide you further.

Thank you for being a part of the Ultralytics community! 🚀

**Reply by Josephts1** [reply]
Here I attach the results obtained using YOLO11s.pt (see Fig.1 and Fig.2), I have frozen the 24 layers of the model to use the pre-trained weights (See code to freeze the layers at the beginning of this discussion). I have used labelme to label my dataset and transformed the .JSON files to .txt (80 images for train, 10 for val and 5 for test)
![image_2025-01-16_142603820](https://github.com/user-attachments/assets/6dc51029-fa14-4882-8a28-8702be7f99ea)
![image_2025-01-16_142652091](https://github.com/user-attachments/assets/e2a7dfbd-1020-4321-8222-cc568e9efd4b)
![image_2025-01-16_142746677](https://github.com/user-attachments/assets/bb71a5e9-b783-4ddf-aec5-287a72dc4a8f)

**Reply by glenn-jocher** [reply]
@Josephts1 thank you for sharing the additional details and results. The sudden variations in metrics could stem from the small dataset size, which may cause instability during training. I recommend increasing the dataset size to improve generalization and stability. Additionally, check data quality to ensure proper labeling and balanced class representation. For further insights, evaluate using the visual outputs mentioned [here](https://docs.ultralytics.com/guides/yolo-performance-metrics/) to analyze performance across thresholds. Let me know if you need clarification!

**Comment by glenn-jocher** [comment, marked as answer]
@Josephts1 thank you for sharing the details. The erratic behavior in your plots is likely due to the small dataset size, which can cause overfitting and high variance in metrics. You may improve stability by augmenting your dataset with more images, using techniques like mosaic or mixup augmentation, and reducing the learning rate or batch size. Also, ensure the dataset is balanced and properly annotated. For further insights, refer to the [train settings guide](https://docs.ultralytics.com/modes/train).
